File: datastructure/array_list/array_list.py
class ArrayList:
    def __init__(self, size=0):
        self.size = size
        self.array_list = [None] * self.size
        self.count = 0

    def __str__(self):
        print([str(item) for item in self.array_list])

    def count(self):  # pylint: disable=E0202
        return self.count + 1

    def is_full(self):
        return self.count + 1 == self.size

    def add(self, item):
        if self.count < self.size:
            self.array_list[self.count] = item
            self.count += 1
            return True
        else:
            return False

    def delete(self, index):
        if index < self.count and index >= 0:
            res = self.array_list[index]
            self.count -= 1
            for i in range(index, self.size):
                if (i == self.count):
                    self.array_list[i] = None
                    return res
                else:
                    self.array_list[i] = self.array_list[i+1]
        else:
            return False


if __name__ == "__main__":
    arr = ArrayList(5)
    arr.add(1)
    arr.add(3)
    arr.add(5)
    arr.add(6)
    arr.add(7)
    arr.delete(1)
    arr.delete(3)
    print([str(item) for item in arr.array_list])

File: datastructure/array_list/array_list.js
class CustomArray {
  constructor(size) {
    this.array = new Array(size).fill(null);
    this.count = 0;
    this.size = size;
  }

  isFull() {
    return this.size == this.count;
  }
  customAdd(item) {
    if (!this.isFull()) {
      this.array[this.count] = item;
      this.count += 1;
      return true;
    } else {
      return false;
    }
  }

  customDelete(index) {
    //check for valid index
    if (index < this.count && index >= 0) {
      // shift left
      for (let i = index; i < this.count - 1; i++) {
        if (i + 1 == this.count - 1) {
          this.array[i + 1] = null;
        } else {
          this.array[i] = this.array[i + 1];
        }
      }
      //decrease count
      this.count -= 1;
      //return element
      return true;
    } else {
      //return false
      return false;
    }
  }
}

let arr = new CustomArray(5);
arr.customAdd(1);
arr.customAdd(2);
arr.customAdd(3);
console.log(arr.array);
arr.customDelete(1);
console.log(arr.array);
